File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if(root.left == None and root.right == None):
            return True

        order = self.inOrder(root, [])

        for i in range(len(order)-1):
            if(order[i] >= order[i+1]):
                return False
        
        return True
    
    def inOrder(self, r, l):
        if(r==None):
            return l
        
        self.inOrder(r.left, l)
        l.append(r.val)
        self.inOrder(r.right, l)
        return l

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return validBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    public boolean validBST(TreeNode r, long low, long hi){
        if(r==null){
            return true;
        }

        if(r.val<=low || r.val>=hi){
            return false;
        }
        boolean left = validBST(r.left, low, r.val);
        boolean right = validBST(r.right, r.val, hi);

        return left && right;
    }
}